zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -1481,4 +1492,307 @@ pub(crate) mod tests {
+
+        // Scan the chain-tip range.
+        st.scan_cached_blocks(birthday.height() + 12, 112);
+
+        // We haven't yet discovered our note, so balances should still be zero
+        assert_eq!(
+            st.get_total_balance(account.account_id()),
+            NonNegativeAmount::ZERO
+        );
+
+        // Now scan the historic range; this should discover our note, which should now be
+        // spendable.
+        st.scan_cached_blocks(birthday.height(), 12);
+        assert_eq!(
+            st.get_total_balance(account.account_id()),
+            NonNegativeAmount::const_from_u64(100000)
+        );
+        assert_eq!(
+            st.get_spendable_balance(account.account_id(), 10),
+            NonNegativeAmount::const_from_u64(100000)
+        );
+
+        // Spend the note.
+        let to_extsk = OrchardPoolTester::sk(&[0xf5; 32]);
+        let to = OrchardPoolTester::sk_default_address(&to_extsk);
+        let request = zip321::TransactionRequest::new(vec![zip321::Payment::without_memo(
+            to.to_zcash_address(&st.network()),
+            NonNegativeAmount::const_from_u64(10000),
+        )])
+        .unwrap();
+
+        let fee_rule = StandardFeeRule::Zip317;
+
+        let change_memo = "Test change memo".parse::<Memo>().unwrap();
+        let change_strategy = standard::SingleOutputChangeStrategy::new(
+            fee_rule,
+            Some(change_memo.into()),
+            OrchardPoolTester::SHIELDED_PROTOCOL,
+        );
+        let input_selector =
+            &GreedyInputSelector::new(change_strategy, DustOutputPolicy::default());
+
+        let proposal = st
+            .propose_transfer(
+                account.account_id(),
+                input_selector,
+                request,
+                NonZeroU32::new(10).unwrap(),
+            )
+            .unwrap();
+
+        let create_proposed_result = st.create_proposed_transactions::<Infallible, _>(
+            account.usk(),
+            OvkPolicy::Sender,
+            &proposal,
+        );
+        assert_matches!(&create_proposed_result, Ok(txids) if txids.len() == 1);
+    }
+
+    #[test]
+    #[cfg(feature = "orchard")]
+    fn orchard_block_spanning_tip_boundary_incomplete() {
+        let mut st = prepare_orchard_block_spanning_test(false);
+        let account = st.test_account().cloned().unwrap();
+        let birthday = account.birthday();
+
+        // set the chain tip to the final position we expect
+        let new_tip = birthday.height() + 123;
+        st.wallet_mut().update_chain_tip(new_tip).unwrap();
+
+        // Verify that the suggested scan ranges includes only the chain-tip range with ChainTip
+        // priority, and that the range from the wallet birthday to the end of the birthday shard
+        // has Historic priority.
+        let birthday_height = birthday.height().into();
+        let expected = vec![
+            scan_range(
+                birthday_height..(new_tip + 1).into(),
```

Comment:
This range not being split in two is clearly due to not updating the subtree roots in `prepare_orchard_block_spanning_test`, but it's not clear to me why this is necessary in addition to skipping the spanning block (to provoke the error), and why the other two possible combinations are not tested. While my earlier comment about documenting `prepare_orchard_block_spanning_test` would help, a comment here (or in `orchard_block_spanning_tip_boundary`, wherever is clearer) would also be useful.

Reply:
This test does not actually reproduce the error; I added it as an initial attempt to reproduce, and then decided that it was useful independently and so decided to keep it. I can split it out of this PR, but it's closely related to the behavior under test (for obvious reasons) so I thought it kind of made sense to keep here?